File: library/src/main/java/com/mikepenz/aboutlibraries/ui/item/HeaderItem.java
package com.mikepenz.aboutlibraries.ui.item;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.mikepenz.aboutlibraries.Libs;
import com.mikepenz.aboutlibraries.LibsBuilder;
import com.mikepenz.aboutlibraries.LibsConfiguration;
import com.mikepenz.aboutlibraries.R;
import com.mikepenz.aboutlibraries.util.MovementCheck;
import com.mikepenz.aboutlibraries.util.RippleForegroundListener;
import com.mikepenz.aboutlibraries.util.UIUtils;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.iconics.Iconics;

import java.util.List;


/**
 * Created by mikepenz on 28.12.15.
 */
public class HeaderItem extends AbstractItem<HeaderItem, HeaderItem.ViewHolder> {
    private Integer aboutVersionCode;
    private String aboutVersionName;
    private Drawable aboutIcon;

    public HeaderItem withAboutVersionCode(Integer aboutVersionCode) {
        this.aboutVersionCode = aboutVersionCode;
        return this;
    }

    public HeaderItem withAboutVersionName(String aboutVersionName) {
        this.aboutVersionName = aboutVersionName;
        return this;
    }

    public HeaderItem withAboutIcon(Drawable aboutIcon) {
        this.aboutIcon = aboutIcon;
        return this;
    }

    public LibsBuilder libsBuilder;

    public HeaderItem withLibsBuilder(LibsBuilder libsBuilder) {
        this.libsBuilder = libsBuilder;
        return this;
    }

    @Override
    public boolean isSelectable() {
        return false;
    }

    /**
     * defines the type defining this item. must be unique. preferably an id
     *
     * @return the type
     */
    @Override
    public int getType() {
        return R.id.header_item_id;
    }

    /**
     * defines the layout which will be used for this item in the list
     *
     * @return the layout for this item
     */
    @Override
    public int getLayoutRes() {
        return R.layout.listheader_opensource;
    }

    /**
     * binds the data of this item onto the viewHolder
     *
     * @param holder the viewHolder of this item
     */
    @Override
    public void bindView(final ViewHolder holder, List<Object> payloads) {
        super.bindView(holder, payloads);

        //ctx
        final Context ctx = holder.itemView.getContext();

        //Set the Icon or hide it
        if (libsBuilder.aboutShowIcon != null && libsBuilder.aboutShowIcon && aboutIcon != null) {
            holder.aboutIcon.setImageDrawable(aboutIcon);
            holder.aboutIcon.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (LibsConfiguration.getInstance().getListener() != null) {
                        LibsConfiguration.getInstance().getListener().onIconClicked(v);
                    }
                }
            });

            holder.aboutIcon.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    return LibsConfiguration.getInstance().getListener() != null && LibsConfiguration.getInstance().getListener().onIconLongClicked(v);

                }
            });
        } else {
            holder.aboutIcon.setVisibility(View.GONE);
        }

        //Set the description or hide it
        if (!TextUtils.isEmpty(libsBuilder.aboutAppName)) {
            holder.aboutAppName.setText(libsBuilder.aboutAppName);
        } else {
            holder.aboutAppName.setVisibility(View.GONE);
        }

        // Reset aboutSpecial fields
        holder.aboutSpecialContainer.setVisibility(View.GONE);
        holder.aboutSpecial1.setVisibility(View.GONE);
        holder.aboutSpecial2.setVisibility(View.GONE);
        holder.aboutSpecial3.setVisibility(View.GONE);

        // set the values for the special fields
        if (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial1) && (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial1Description) || LibsConfiguration.getInstance().getListener() != null)) {
            holder.aboutSpecial1.setText(libsBuilder.aboutAppSpecial1);
            new Iconics.IconicsBuilder().ctx(ctx).on(holder.aboutSpecial1).build();
            holder.aboutSpecial1.setVisibility(View.VISIBLE);
            holder.aboutSpecial1.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    boolean consumed = false;
                    if (LibsConfiguration.getInstance().getListener() != null) {
                        consumed = LibsConfiguration.getInstance().getListener().onExtraClicked(v, Libs.SpecialButton.SPECIAL1);
                    }

                    if (!consumed && !TextUtils.isEmpty(libsBuilder.aboutAppSpecial1Description)) {
                        try {
                            AlertDialog alert = new AlertDialog.Builder(ctx)
                                .setMessage(Html.fromHtml(libsBuilder.aboutAppSpecial1Description))
                                .create();
                            alert.show();
                            TextView alertText = (TextView) alert.findViewById(android.R.id.message);
                            if (alertText != null) {
                                alertText.setMovementMethod(LinkMovementMethod.getInstance());
                            }
                        } catch (Exception ex) {
                        }
                    }
                }
            });
            holder.aboutSpecialContainer.setVisibility(View.VISIBLE);
        }
        if (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial2) && (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial2Description) || LibsConfiguration.getInstance().getListener() != null)) {
            holder.aboutSpecial2.setText(libsBuilder.aboutAppSpecial2);
            new Iconics.IconicsBuilder().ctx(ctx).on(holder.aboutSpecial2).build();
            holder.aboutSpecial2.setVisibility(View.VISIBLE);
            holder.aboutSpecial2.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    boolean consumed = false;
                    if (LibsConfiguration.getInstance().getListener() != null) {
                        consumed = LibsConfiguration.getInstance().getListener().onExtraClicked(v, Libs.SpecialButton.SPECIAL2);
                    }

                    if (!consumed && !TextUtils.isEmpty(libsBuilder.aboutAppSpecial2Description)) {
                        try {
                            AlertDialog alert = new AlertDialog.Builder(ctx)
                                .setMessage(Html.fromHtml(libsBuilder.aboutAppSpecial2Description))
                                .create();
                            alert.show();
                            TextView alertText = (TextView) alert.findViewById(android.R.id.message);
                            if (alertText != null) {
                                alertText.setMovementMethod(LinkMovementMethod.getInstance());
                            }
                        } catch (Exception ex) {
                        }
                    }
                }
            });
            holder.aboutSpecialContainer.setVisibility(View.VISIBLE);
        }
        if (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial3) && (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial3Description) || LibsConfiguration.getInstance().getListener() != null)) {
            holder.aboutSpecial3.setText(libsBuilder.aboutAppSpecial3);
            new Iconics.IconicsBuilder().ctx(ctx).on(holder.aboutSpecial3).build();
            holder.aboutSpecial3.setVisibility(View.VISIBLE);
            holder.aboutSpecial3.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    boolean consumed = false;
                    if (LibsConfiguration.getInstance().getListener() != null) {
                        consumed = LibsConfiguration.getInstance().getListener().onExtraClicked(v, Libs.SpecialButton.SPECIAL3);
                    }

                    if (!consumed && !TextUtils.isEmpty(libsBuilder.aboutAppSpecial3Description)) {
                        try {
                            AlertDialog alert = new AlertDialog.Builder(ctx)
                                .setMessage(Html.fromHtml(libsBuilder.aboutAppSpecial3Description))
                                .create();
                            alert.show();
                            TextView alertText = (TextView) alert.findViewById(android.R.id.message);
                            if (alertText != null) {
                                alertText.setMovementMethod(LinkMovementMethod.getInstance());
                            }
                        } catch (Exception ex) {
                        }
                    }
                }
            });
            holder.aboutSpecialContainer.setVisibility(View.VISIBLE);
        }


        //set the Version or hide it
        if (libsBuilder.aboutVersionString != null)
            holder.aboutVersion.setText(libsBuilder.aboutVersionString);
        else {
            if (libsBuilder.aboutShowVersion != null && libsBuilder.aboutShowVersion) {
                holder.aboutVersion.setText(ctx.getString(R.string.version) + " " + aboutVersionName + " (" + aboutVersionCode + ")");
            } else {
                if (libsBuilder.aboutShowVersionName != null && libsBuilder.aboutShowVersionName) {
                    holder.aboutVersion.setText(ctx.getString(R.string.version) + " " + aboutVersionName);
                } else if (libsBuilder.aboutShowVersionCode != null && libsBuilder.aboutShowVersionCode) {
                    holder.aboutVersion.setText(ctx.getString(R.string.version) + " " + aboutVersionCode);
                } else {
                    holder.aboutVersion.setVisibility(View.GONE);
                }
            }
        }

        //Set the description or hide it
        if (!TextUtils.isEmpty(libsBuilder.aboutDescription)) {
            holder.aboutAppDescription.setText(Html.fromHtml(libsBuilder.aboutDescription));
            new Iconics.IconicsBuilder().ctx(ctx).on(holder.aboutAppDescription).build();
            holder.aboutAppDescription.setMovementMethod(MovementCheck.getInstance());
        } else {
            holder.aboutAppDescription.setVisibility(View.GONE);
        }

        //if there is no description or no icon and version number hide the divider
        if (!libsBuilder.aboutShowIcon && !libsBuilder.aboutShowVersion || TextUtils.isEmpty(libsBuilder.aboutDescription)) {
            holder.aboutDivider.setVisibility(View.GONE);
        }

        //notify the libsRecyclerViewListener to allow modifications
        if (LibsConfiguration.getInstance().getLibsRecyclerViewListener() != null) {
            LibsConfiguration.getInstance().getLibsRecyclerViewListener().onBindViewHolder(holder);
        }
    }

    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    /**
     * our ViewHolder
     */
    public static class ViewHolder extends RecyclerView.ViewHolder {
        ImageView aboutIcon;
        TextView aboutAppName;
        View aboutSpecialContainer;
        Button aboutSpecial1;
        Button aboutSpecial2;
        Button aboutSpecial3;
        TextView aboutVersion;
        View aboutDivider;
        TextView aboutAppDescription;

        public ViewHolder(View headerView) {
            super(headerView);

            //get the about this app views
            aboutIcon = (ImageView) headerView.findViewById(R.id.aboutIcon);
            aboutAppName = (TextView) headerView.findViewById(R.id.aboutName);
            aboutAppName.setTextColor(UIUtils.getThemeColorFromAttrOrRes(headerView.getContext(), R.attr.about_libraries_title_description, R.color.about_libraries_title_description));
            aboutSpecialContainer = headerView.findViewById(R.id.aboutSpecialContainer);
            aboutSpecial1 = (Button) headerView.findViewById(R.id.aboutSpecial1);
            aboutSpecial2 = (Button) headerView.findViewById(R.id.aboutSpecial2);
            aboutSpecial3 = (Button) headerView.findViewById(R.id.aboutSpecial3);
            aboutVersion = (TextView) headerView.findViewById(R.id.aboutVersion);
            aboutVersion.setTextColor(UIUtils.getThemeColorFromAttrOrRes(headerView.getContext(), R.attr.about_libraries_text_description, R.color.about_libraries_text_description));
            aboutDivider = headerView.findViewById(R.id.aboutDivider);
            aboutDivider.setBackgroundColor(UIUtils.getThemeColorFromAttrOrRes(headerView.getContext(), R.attr.about_libraries_divider_description, R.color.about_libraries_divider_description));
            aboutAppDescription = (TextView) headerView.findViewById(R.id.aboutDescription);
            aboutAppDescription.setTextColor(UIUtils.getThemeColorFromAttrOrRes(headerView.getContext(), R.attr.about_libraries_text_description, R.color.about_libraries_text_description));
        }
    }
}

File: library/src/main/java/com/mikepenz/aboutlibraries/ui/item/HeaderItem.kt
package com.mikepenz.aboutlibraries.ui.item

import android.graphics.drawable.Drawable
import android.text.Html
import android.text.TextUtils
import android.text.method.LinkMovementMethod
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.aboutlibraries.Libs
import com.mikepenz.aboutlibraries.LibsBuilder
import com.mikepenz.aboutlibraries.LibsConfiguration
import com.mikepenz.aboutlibraries.R
import com.mikepenz.aboutlibraries.util.MovementCheck
import com.mikepenz.aboutlibraries.util.getSupportColor
import com.mikepenz.aboutlibraries.util.getThemeColor
import com.mikepenz.aboutlibraries.util.resolveStyledValue
import com.mikepenz.fastadapter.items.AbstractItem

/**
 * Created by mikepenz on 28.12.15.
 */
class HeaderItem(var libsBuilder: LibsBuilder) : AbstractItem<HeaderItem.ViewHolder>() {
    private var aboutVersionCode: Int? = null
    private var aboutVersionName: String? = null
    private var aboutIcon: Drawable? = null

    fun withAboutVersionCode(aboutVersionCode: Int?): HeaderItem {
        this.aboutVersionCode = aboutVersionCode
        return this
    }

    fun withAboutVersionName(aboutVersionName: String?): HeaderItem {
        this.aboutVersionName = aboutVersionName
        return this
    }

    fun withAboutIcon(aboutIcon: Drawable?): HeaderItem {
        this.aboutIcon = aboutIcon
        return this
    }

    fun withLibsBuilder(libsBuilder: LibsBuilder): HeaderItem {
        this.libsBuilder = libsBuilder
        return this
    }

    override var isSelectable: Boolean
        get() = false
        set(value) {}

    /**
     * defines the type defining this item. must be unique. preferably an id
     *
     * @return the type
     */
    override val type: Int
        get() = R.id.header_item_id

    /**
     * defines the layout which will be used for this item in the list
     *
     * @return the layout for this item
     */
    override val layoutRes: Int
        get() = R.layout.listheader_opensource

    /**
     * binds the data of this item onto the viewHolder
     *
     * @param holder the viewHolder of this item
     */
    override fun bindView(holder: ViewHolder, payloads: List<Any>) {
        super.bindView(holder, payloads)

        //ctx
        val ctx = holder.itemView.context

        //Set the Icon or hide it
        if (libsBuilder.aboutShowIcon && aboutIcon != null) {
            holder.aboutIcon.setImageDrawable(aboutIcon)
            holder.aboutIcon.setOnClickListener {
                LibsConfiguration.listener?.onIconClicked(it)
            }

            holder.aboutIcon.setOnLongClickListener { v -> LibsConfiguration.listener != null && LibsConfiguration.listener?.onIconLongClicked(v) ?: false }
        } else {
            holder.aboutIcon.visibility = View.GONE
        }

        //Set the description or hide it
        if (!libsBuilder.aboutAppName.isNullOrEmpty()) {
            holder.aboutAppName.text = libsBuilder.aboutAppName
        } else {
            holder.aboutAppName.visibility = View.GONE
        }

        // Reset aboutSpecial fields
        holder.aboutSpecialContainer.visibility = View.GONE
        holder.aboutSpecial1.visibility = View.GONE
        holder.aboutSpecial2.visibility = View.GONE
        holder.aboutSpecial3.visibility = View.GONE

        // set the values for the special fields
        if (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial1) && (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial1Description) || LibsConfiguration.listener != null)) {
            holder.aboutSpecial1.text = libsBuilder.aboutAppSpecial1
            LibsConfiguration.postTextAction?.invoke(holder.aboutSpecial1)
            holder.aboutSpecial1.visibility = View.VISIBLE
            holder.aboutSpecial1.setOnClickListener { v ->
                val consumed = LibsConfiguration.listener?.onExtraClicked(v, Libs.SpecialButton.SPECIAL1)
                        ?: false

                if (!consumed && !TextUtils.isEmpty(libsBuilder.aboutAppSpecial1Description)) {
                    try {
                        val alert = AlertDialog.Builder(ctx)
                                .setMessage(Html.fromHtml(libsBuilder.aboutAppSpecial1Description))
                                .create()
                        alert.show()
                        val alertText = alert.findViewById<View>(android.R.id.message) as TextView?
                        if (alertText != null) {
                            alertText.movementMethod = LinkMovementMethod.getInstance()
                        }
                    } catch (ignored: Exception) {
                        // ignored
                    }
                }
            }
            holder.aboutSpecialContainer.visibility = View.VISIBLE
        }
        if (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial2) && (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial2Description) || LibsConfiguration.listener != null)) {
            holder.aboutSpecial2.text = libsBuilder.aboutAppSpecial2
            LibsConfiguration.postTextAction?.invoke(holder.aboutSpecial2)
            holder.aboutSpecial2.visibility = View.VISIBLE
            holder.aboutSpecial2.setOnClickListener { v ->
                val consumed = LibsConfiguration.listener?.onExtraClicked(v, Libs.SpecialButton.SPECIAL2)
                        ?: false
                if (!consumed && !TextUtils.isEmpty(libsBuilder.aboutAppSpecial2Description)) {
                    try {
                        val alert = AlertDialog.Builder(ctx)
                                .setMessage(Html.fromHtml(libsBuilder.aboutAppSpecial2Description))
                                .create()
                        alert.show()
                        val alertText = alert.findViewById<View>(android.R.id.message) as TextView?
                        if (alertText != null) {
                            alertText.movementMethod = LinkMovementMethod.getInstance()
                        }
                    } catch (ignored: Exception) {
                        // ignored
                    }
                }
            }
            holder.aboutSpecialContainer.visibility = View.VISIBLE
        }
        if (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial3) && (!TextUtils.isEmpty(libsBuilder.aboutAppSpecial3Description) || LibsConfiguration.listener != null)) {
            holder.aboutSpecial3.text = libsBuilder.aboutAppSpecial3
            LibsConfiguration.postTextAction?.invoke(holder.aboutSpecial3)
            holder.aboutSpecial3.visibility = View.VISIBLE
            holder.aboutSpecial3.setOnClickListener { v ->
                val consumed = LibsConfiguration.listener?.onExtraClicked(v, Libs.SpecialButton.SPECIAL3)
                        ?: false

                if (!consumed && !TextUtils.isEmpty(libsBuilder.aboutAppSpecial3Description)) {
                    try {
                        val alert = AlertDialog.Builder(ctx)
                                .setMessage(Html.fromHtml(libsBuilder.aboutAppSpecial3Description))
                                .create()
                        alert.show()
                        val alertText = alert.findViewById<View>(android.R.id.message) as TextView?
                        if (alertText != null) {
                            alertText.movementMethod = LinkMovementMethod.getInstance()
                        }
                    } catch (ignored: Exception) {
                        // ignored
                    }
                }
            }
            holder.aboutSpecialContainer.visibility = View.VISIBLE
        }

        //set the Version or hide it
        if (libsBuilder.aboutVersionString.isNotEmpty())
            holder.aboutVersion.text = libsBuilder.aboutVersionString
        else {
            if (libsBuilder.aboutShowVersion) {
                holder.aboutVersion.text = "${ctx.getString(R.string.version)} $aboutVersionName ($aboutVersionCode)"
            } else {
                when {
                    libsBuilder.aboutShowVersionName -> holder.aboutVersion.text = "${ctx.getString(R.string.version)} $aboutVersionName"
                    libsBuilder.aboutShowVersionCode -> holder.aboutVersion.text = "${ctx.getString(R.string.version)} $aboutVersionCode"
                    else -> holder.aboutVersion.visibility = View.GONE
                }
            }
        }

        //Set the description or hide it
        if (!libsBuilder.aboutDescription.isNullOrEmpty()) {
            holder.aboutAppDescription.text = Html.fromHtml(libsBuilder.aboutDescription)
            LibsConfiguration.postTextAction?.invoke(holder.aboutAppDescription)
            holder.aboutAppDescription.movementMethod = MovementCheck.instance
        } else {
            holder.aboutAppDescription.visibility = View.GONE
        }

        //if there is no description or no icon and version number hide the divider
        if ((!libsBuilder.aboutShowIcon) && (!libsBuilder.aboutShowVersion) || TextUtils.isEmpty(libsBuilder.aboutDescription)) {
            holder.aboutDivider.visibility = View.GONE
        }

        //notify the libsRecyclerViewListener to allow modifications
        LibsConfiguration.libsRecyclerViewListener?.onBindViewHolder(holder)
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    /**
     * our ViewHolder
     */
    class ViewHolder(headerView: View) : RecyclerView.ViewHolder(headerView) {
        internal var aboutIcon: ImageView = headerView.findViewById(R.id.aboutIcon) as ImageView
        internal var aboutAppName: TextView = headerView.findViewById(R.id.aboutName) as TextView
        internal var aboutSpecialContainer: View = headerView.findViewById(R.id.aboutSpecialContainer)
        internal var aboutSpecial1: Button = headerView.findViewById(R.id.aboutSpecial1) as Button
        internal var aboutSpecial2: Button = headerView.findViewById(R.id.aboutSpecial2) as Button
        internal var aboutSpecial3: Button = headerView.findViewById(R.id.aboutSpecial3) as Button
        internal var aboutVersion: TextView = headerView.findViewById(R.id.aboutVersion) as TextView
        internal var aboutDivider: View = headerView.findViewById(R.id.aboutDivider)
        internal var aboutAppDescription: TextView = headerView.findViewById(R.id.aboutDescription) as TextView

        init {
            val ctx = itemView.context
            ctx.resolveStyledValue {
                aboutAppName.setTextColor(it.getColorStateList(R.styleable.AboutLibraries_aboutLibrariesDescriptionTitle))
                aboutVersion.setTextColor(it.getColorStateList(R.styleable.AboutLibraries_aboutLibrariesDescriptionText))
                aboutAppDescription.setTextColor(it.getColorStateList(R.styleable.AboutLibraries_aboutLibrariesDescriptionText))
                aboutDivider.setBackgroundColor(it.getColor(R.styleable.AboutLibraries_aboutLibrariesDescriptionDivider, ctx.getThemeColor(R.attr.aboutLibrariesDescriptionDivider, ctx.getSupportColor(R.color.about_libraries_dividerLight_openSource))))
                aboutSpecial1.setTextColor(it.getColorStateList(R.styleable.AboutLibraries_aboutLibrariesSpecialButtonText))
                aboutSpecial2.setTextColor(it.getColorStateList(R.styleable.AboutLibraries_aboutLibrariesSpecialButtonText))
                aboutSpecial3.setTextColor(it.getColorStateList(R.styleable.AboutLibraries_aboutLibrariesSpecialButtonText))
            }
        }
    }
}